### DISCUSSION TITLE
get/set Expression bodied properties

### DISCUSSION THREAD

**Discussion by MkazemAkhgary** [dotnet/csharplang#1007] [General]
Allow to get and set with expression Bodied properties

```
public int X <=> point.X;
```
Equivalent to

```
public int X
{
    get => point.X;
    set => point.X = value;
}
```

Could be used for indexers too.

```
public int this[int index] <=> _source[index]; 
```

Equivalent to 

```
public int this[int index]
{
    get => _source[index];
    set => _source[index] = value;
}
```

----

Notes to consider 

 - expression must be a valid l-value **and** r-value (as suggested by @benaadams)

**Comment by Logerfo** [comment]
>`int Foo <=> bar = baz = qux; // sets all values.`

and returns...?

**Comment by DavidArno** [comment]
I initially upvoted this as a simple idea for mapping read/write properties to some value other than the default backing field. But I've removed that upvote as your "*Notes to consider*" addendum just muddies the proposal.

Keep it simple:
1. For property, `x <=> item`, `item` can be a (non-readonly) field or a read/write property,
2. For indexer, `this[int index] <=> item`, `item` must be another indexer.

**Comment by benaadams** [comment]
@DavidArno More explicit on the indexer?
```
this[int index] <=> item[index];
this[int index] <=> item[_offset + index];
```
i.e. must always resolve to a single value/ref

**Comment by MkazemAkhgary** [comment]
@Logerfo returns what `=> x = y = z;` returns. 

@DavidArno indeed, i tried to expand usage in conjunction with get only properties. Chained assignments could be prevented. Its not something i do usually, in fact i never did it.

**Comment by MkazemAkhgary** [comment]
@benaadams i liked that but what if item is jagged array `int[][]`?

Then it has to be either explicit like what you suggest or implicit `item`.

I guess explicit version pays off more.

**Comment by Logerfo** [comment]
@MkazemAkhgary in C#, multiple assignments returns the rightmost value of the chain. In your examples, `qux` or `z`, but that can be really confusing. I see you already remove that from the notes, so we don't need to proceed with this discussion.

**Comment by benaadams** [comment]
>  i liked that but what if item is jagged array int[][]?

Defined via type, an array is still a single item
```csharp
private int[][] item;

public int[] this[int index] <=> item[index];
```
or
```csharp
private int[][] item;

public int this[int i1, int i2] <=> item[i1][i2];
```
or the other type
```csharp
private int[,] item;

public int this[int i1, int i2] <=> item[i1, i2];
```

**Comment by benaadams** [comment]
I think it just needs to be

"expression must be a valid l-value and r-value"

i.e. valid on both sides of an equals

For example it could happily be a `ref` returning function so for example given
```csharp
private int[] items;

private ref int FindEntry(int index)
{
    // ...
}
```
This
```csharp
public int X
{
    get => FindEntry(0);
    set => FindEntry(0) = value;
}

public int this[int index]
{
    get => FindEntry(index);
    set => FindEntry(index) = value;
}
```
Could be converted to
```csharp
public int X <=> FindEntry(0);

public int this[int index] <=> FindEntry(index);
```

**Comment by Logerfo** [comment]
Another possibility is to use just the `=>` syntax and change its behavior to act like you propose, now allowing the `readonly` modifier on properties to behave the way the operator does today. That would change existing properties, but I don't know how big deal that would be.

**Comment by benaadams** [comment]
> That would change existing properties, but I don't know how big deal that would be.

Would be a significant breaking change as all current simple readonly properties would suddenly become read/write e.g.

```csharp
private int _x;

public int X => _x; // now read/write, was readonly before
```

**Comment by Logerfo** [comment]
@benaadams Since the assignment is not possible today, that change would not actually break anything, as it's more kind of a relaxation. It can break analyzers, but anything can. I think it would only break reflection/cecil for properties like `IsWriteable` or `IsReadonly`, which would have different values due to this change.

**Comment by MkazemAkhgary** [comment]
@Logerfo not all properties can become writable. issue is they will remain get only or become writable.  so whether property will be readonly or not would become vague. I guess `=>` should be left as it is for get only properties.

for example  `X => 1 + 2` is impossible to be writable.

**Comment by benaadams** [comment]
@Logerfo hmm true; but it would also mean everyone who had written a readonly property would then have to change/audit their code to keep the properties readonly

**Comment by MkazemAkhgary** [comment]
@benaadams very good explanation. that is more general, updated.

**Comment by ghost** [comment]
I suggest `using` instead of `<=>`:
```C#
public int this[int index] using _source[index]; 
```

#2291

**Comment by Happypig375** [comment]
That syntax deviates too much from `=>`. It looks too different from `=>`.

**Comment by heku** [comment]
Maybe extend the `ref`'s usage is better:
```c#
public int this[int index] ref _source[index]; 
public int Prop ref _field;
```

**Comment by TahirAhmadov** [comment]
I know the `<=>` syntax is very short, but I was thinking a slightly different syntax which is a bit more verbose but I think easier to read:
```cs
public string Prop { get, set => this._field; }
```

**Reply by FaustVX** [reply]
With proposal #140 the syntax `public string Prop { get; set => this._field; }` but with a `;` after the `get`, it will be a too much close syntax.

**Reply by TahirAhmadov** [reply]
Good catch. I take it back.

**Reply by sab39** [reply]
Personally I think `<=>` wins out on readability anyway: since we're all already used to `=>` as a "one way" mapping, adding the other arrow intuitively means the same thing but bidirectional.

**Comment by LWChris** [comment]
## Alternative 1

`T P <=> X;`

where
* `T` is any type
* `P` is a property name
* `X` is an expression for a readable and writable property or field of type `T`

## Alternative 2

`T M <=> X;`
* `T` is any type
* `M` is a property name **_or indexer_**
* `X` is an expression for a readable and writable property, **_indexer_** or field of type `T`

---

I'm in full support of alternative 1, and somewhat supportive of alternative 2.

I'm per se not a fan of "complex" expressions on the RHS like

`T M <=> x = y;`

I know it's already possible to write

```
T M {
    get => x = y;
    set => x = y = value;
}
```

but I am a bit worried that people might read the shorthand syntax as

`T M <=> x    = y;`

and not realize that `get` will have a side effect on `x` and `set` will overwrite `y`. Therefore I'm leaning towards strict limitations what `X` can be if possible.

Indexer expressions are on the boundary of "complex" RHS expressions, because they take additional arguments. But seeing that indexer delegation is probably often required alongside property delegation, if the wrapped/delegated object happens to have an indexer property, I'm slightly less worried about that.

**Reply by LWChris** [reply]
As for the syntax, I'm heavily in favor for `<=>` because it is the natural instinct, when tasked "make `=>` bi-directional", to make it `<=>`.

I know multiple people, especially @CyrusNajmabadi wants to have `<=>` (aka "spaceship operator") for three-way-comparison (result is not `bool` but something that can be "greater than, less than, equal"), but for that I'd prefer `<>=`, precisely to _avoid_ making the expression `x <=> y` look like an arrow opposed to `x <>= y`.
